File: py/horus/rpc/services.py
from dataclasses import dataclass as _dataclass
from typing import Optional as _Optional

@_dataclass(frozen=True)
class RpcServiceInfo:
    """Information about a Horus RPC service."""

    full_name: str
    "Unique full name of the service, e.g. `horus.pb.NotificationService`."
    id: int
    "Unique identifier of the service, e.g. 1."
    host: str
    "Host to connect to the service, e.g. `localhost`."
    port: int
    "Port to connect to the service, e.g. `40001`."

    @property
    def url(self) -> str:
        """URL used to connect to the service."""
        return f"ws://{self.host}:{self.port}"

    def with_(self, /, host: _Optional[str]=None, port: _Optional[int]=None) -> "RpcServiceInfo":
        """Returns a copy of `self` with `host` and `port` replaced by the given values."""
        return RpcServiceInfo(
            full_name=self.full_name,
            id=self.id,
            host=host if host is not None else self.host,
            port=port if port is not None else self.port,
        )

@_dataclass(frozen=True)
class RpcServices:
    """Enumeration of the Horus RPC services."""

    detection_merger: RpcServiceInfo = RpcServiceInfo(
        full_name="horus.pb.DetectionMergerService",
        id=11,
        host="127.0.0.1",
        port=40011,
    )

    detection: RpcServiceInfo = RpcServiceInfo(
        full_name="horus.pb.DetectionService",
        id=7,
        host="127.0.0.1",
        port=40007,
    )

    notification: RpcServiceInfo = RpcServiceInfo(
        full_name="horus.pb.NotificationService",
        id=2,
        host="127.0.0.1",
        port=40002,
    )

    point_aggregator: RpcServiceInfo = RpcServiceInfo(
        full_name="horus.pb.PointAggregatorService",
        id=6,
        host="127.0.0.1",
        port=40006,
    )

    project_manager: RpcServiceInfo = RpcServiceInfo(
        full_name="horus.pb.ProjectManagerService",
        id=3,
        host="127.0.0.1",
        port=40003,
    )

    status: RpcServiceInfo = RpcServiceInfo(
        full_name="horus.pb.StatusService",
        id=15,
        host="127.0.0.1",
        port=40015,
    )

    def with_(self,
              /,
              detection_merger: _Optional[RpcServiceInfo]=None,
              detection: _Optional[RpcServiceInfo]=None,
              notification: _Optional[RpcServiceInfo]=None,
              point_aggregator: _Optional[RpcServiceInfo]=None,
              project_manager: _Optional[RpcServiceInfo]=None,
              status: _Optional[RpcServiceInfo]=None,
             ) -> "RpcServices":
        """Returns a copy of `self` with services replaced as specified."""
        return RpcServices(
            detection_merger=detection_merger or self.detection_merger,
            detection=detection or self.detection,
            notification=notification or self.notification,
            point_aggregator=point_aggregator or self.point_aggregator,
            project_manager=project_manager or self.project_manager,
            status=status or self.status,
        )

File: go/horus/services.go
package horus

// ServiceInfo provides information about a Horus RPC service.
type ServiceInfo struct {
	// Unique full name of the service, e.g. `horus.pb.NotificationService`.
	FullName string
	// Unique identifier of the service, e.g. 1.
	Id uint16
	// Host to connect to the service, e.g. `localhost`.
	Host string
	// Port to connect to the service, e.g. `40001`.
	Port uint16
}

// WithHost returns a copy of si with its [ServiceInfo.Host] set to host.
func (si ServiceInfo) WithHost(host string) ServiceInfo {
	si.Host = host
	return si
}

// WithPort returns a copy of si with its [ServiceInfo.Port] set to port.
func (si ServiceInfo) WithPort(port uint16) ServiceInfo {
	si.Port = port
	return si
}

// Services provides the [ServiceInfo] of the Horus RPC services.
type Services struct {
	Detection ServiceInfo
	Notification ServiceInfo
	PointAggregator ServiceInfo
	ProjectManager ServiceInfo
}

// WithDetection returns a copy of s with its [Services.Detection] set to si.
func (s Services) WithDetection(si ServiceInfo) Services {
	s.Detection = si
	return s
}

// WithNotification returns a copy of s with its [Services.Notification] set to si.
func (s Services) WithNotification(si ServiceInfo) Services {
	s.Notification = si
	return s
}

// WithPointAggregator returns a copy of s with its [Services.PointAggregator] set to si.
func (s Services) WithPointAggregator(si ServiceInfo) Services {
	s.PointAggregator = si
	return s
}

// WithProjectManager returns a copy of s with its [Services.ProjectManager] set to si.
func (s Services) WithProjectManager(si ServiceInfo) Services {
	s.ProjectManager = si
	return s
}

// DefaultServices returns the default configuration of [Services].
func DefaultServices() Services {
	return Services{
		Detection: ServiceInfo{
			FullName: "horus.pb.DetectionService",
			Id: 7,
			Host: "127.0.0.1",
			Port: 40007,
		},
		Notification: ServiceInfo{
			FullName: "horus.pb.NotificationService",
			Id: 2,
			Host: "127.0.0.1",
			Port: 40002,
		},
		PointAggregator: ServiceInfo{
			FullName: "horus.pb.PointAggregatorService",
			Id: 6,
			Host: "127.0.0.1",
			Port: 40006,
		},
		ProjectManager: ServiceInfo{
			FullName: "horus.pb.ProjectManagerService",
			Id: 3,
			Host: "127.0.0.1",
			Port: 40003,
		},
	}
}
